### PR TITLE
Add itertools.islice support in dynamo

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133820

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/polyfill.py [hunk 1]
```diff
@@ -24,12 +24,33 @@ def any(iterator):
 
 
 def index(iterator, item, start=0, end=None):
-    for i, elem in list(enumerate(iterator))[start:end]:
+    for i, elem in islice(enumerate(list(iterator)), start, end):
```
Comment (Skylion007) [inline torch/_dynamo/polyfill.py:27]: Ugh, enumerate and list should operate direclty on iterators, why is the list cast needed?
```suggestion
    for i, elem in islice(enumerate(iterator), start, end):
```

File: torch/_dynamo/polyfill.py [hunk 2]
```diff
@@ -24,13 +24,36 @@ def any(iterator):
...
 def index(iterator, item, start=0, end=None):
-    for i, elem in list(enumerate(iterator))[start:end]:
+    for i, elem in islice(enumerate(iterator), start, end):
         if item == elem:
             return i
     # This will not run in dynamo
     raise ValueError(f"{item} is not in {type(iterator)}")
 
 
+def islice(iterator, start=0, end=None, step=1):
```
Comment (oulgen) [inline torch/_dynamo/polyfill.py:34]: why did you need to reimplement islice?

File: torch/_dynamo/variables/iter.py [hunk 3]
```diff
@@ -174,6 +174,12 @@ def retrieve_const_key(key):
                     from_exc=e,
                 )
             return variables.ListIteratorVariable(result, mutable_local=MutableLocal())
+        elif self.value is itertools.islice:
+            from .builder import SourcelessBuilder
+
+            return tx.inline_user_function_return(
+                SourcelessBuilder.create(tx, polyfill.islice), args, kwargs
```
Comment (oulgen) [inline torch/_dynamo/variables/iter.py:181]: ah this one just calls the implementation in polyfill

File: torch/_dynamo/polyfill.py (hunk 2 above)
Comment (oulgen) [reply torch/_dynamo/polyfill.py:34]: read the dynamo code to understand this, looks good!

**Review by oulgen — APPROVED** [review]

**Comment by oulgen** [comment]
For future reference, we merge using `@pytorchbot merge` instead of doing it manually, i'm actually quite surprised the manual merge worked.

**Comment by oulgen** [comment]
@bobrenjc93 looks like you merged it into your branch, i'm merging it to origin/main now
